### DISCUSSION TITLE
Is there a way to get back the updated hash for any given file in the PDS?

### DISCUSSION THREAD

**Discussion by jordanpadams** [NASA-PDS/pds-api#258] [Q&A] [answered]
Is there a plan for a "service" from the PDS to deliver updated hash registries for any given file?

**Comment by pds-ops** [comment, marked as answer]
This is already possible through the PDS Search API for "most" PDS4 data (working on loading all...):
```
$ curl --get 'https://pds.nasa.gov/api/search/1/products/urn:nasa:pds:insight_rad:data_raw:hp3_rad_raw_00004_20181130_085325' | json_pp

...
{
   "lidvid" : "urn:nasa:pds:insight_rad:data_raw:hp3_rad_raw_00004_20181130_085325::3.1",
...
      "ops:Label_File_Info.ops:md5_checksum" : [
         "985cc9d4e731e5e6e3b22e64b0f62bdf"
      ],
...
      "ops:Data_File_Info.ops:md5_checksum" : [
         "9be7810ae4f744af51eb6e6d6755e11a"
      ],
...
}
```

or here is how you can request just the checksums:
```
$ curl --get 'https://pds.nasa.gov/api/search/1/products/urn:nasa:pds:insight_rad:data_raw:hp3_rad_raw_00004_20181130_085325?fields=lidvid,ops:Label_File_Info.ops:md5_checksum,ops:Data_File_Info.ops:md5_checksum' \
   --header 'Accept: application/csv'
```

More info here: https://nasa-pds.github.io/pds-api/guides/search/endpoints.html

if there are specific data sets you think are missing or you need for your efforts, let us know

we will prioritize those ASAP. Python API here: https://nasa-pds.github.io/pds-api-client/

the API returns only the latest version by default. for past versions: https://pds.nasa.gov/api/search/1/products/{lidvid}/all


to check if a newer version of a product exists, you can query by LID, and check the version_id or creation_date_time. for example, https://pds.nasa.gov/api/search/1/products/urn:nasa:pds:insight.spice:spice_kernels, will return the latest version but https://pds.nasa.gov/api/search/1/products/urn:nasa:pds:insight.spice:spice_kernels/all returns all 16 versions so you can cache the version_id or creation_date_time that comes back from https://pds.nasa.gov/api/search/1/products/urn:nasa:pds:insight.spice:spice_kernels and ping that endpoint regularly to determine if it has changed:
```
{
  ...
  "properties" : {
    "vid" : [ "16.0" ],
    ...
    "pds:Identification_Area.pds:version_id" : [ "16.0" ],
   ...
       "ops:Label_File_Info.ops:creation_date_time" : [ "2023-03-20T19:42:30Z" ],
   ...
       "ops:Data_File_Info.ops:creation_date_time" : [ "2023-03-20T19:42:30Z" ],
   ...
}
```

if you want to actually track down the files on the file system, a couple more lines of code and you can figure that out:
```
"ops:Label_File_Info.ops:file_ref" : [ "https://naif.jpl.nasa.gov/pub/naif/pds/pds4/insight/insight_spice/spice_kernels/collection_spice_kernels_v016.xml" ],

"ops:Data_File_Info.ops:file_ref" : [ "https://naif.jpl.nasa.gov/pub/naif/pds/pds4/insight/insight_spice/spice_kernels/collection_spice_kernels_inventory_v016.csv" ],
```
